Add DynamicProvisioningScheduling support for GCE PD and RePD

ddebroy opened the pull request:
**What this PR does / why we need it**:
This PR adds support for the DynamicProvisioningScheduling feature for GCE PD and RePD. With this in place, if VolumeBindingMode: WaitForFirstConsumer is specified in a GCE storageclass and DynamicProvisioningScheduling is enabled, GCE PD provisioner will use the selected node's LabelZoneFailureDomain as (1) the zone to provision a GCE PD volume in (2) one of the zones to provision GCE RePD volume in.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
E2E tests for DynamicProvisioningScheduling scenarios for GCE PD to follow 

**Release note**:
```release-note
none
```
/sig storage
/assign @msau42

idealhack commented:
/ok-to-test
/assign @saad-ali

ddebroy commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

msau42 commented on pkg/volume/gce_pd/gce_util.go:
```diff
@@ -148,61 +147,45 @@ func (gceutil *GCEDiskUtil) CreateVolume(c *gcePersistentDiskProvisioner) (strin
+	var activezones sets.String
+	activezones, err = cloud.GetAllCurrentZones()
+	if err != nil {
+		glog.V(2).Infof("error getting zone information from GCE: %v", err)
```
does this need to be logged since it will be returned as an error?

ddebroy replied:
This was in the original code but unnecessary. Will remove.

msau42 commented on pkg/volume/util/util_test.go:
```diff
@@ -1460,6 +1460,612 @@ func TestSelectZoneForVolume(t *testing.T) {
+				},
+			},
+			Reject: true,
+		},
```
Another negative test case could be not enough active zones

msau42 commented on pkg/volume/util/util_test.go:
```diff
@@ -1460,6 +1460,612 @@ func TestSelectZoneForVolume(t *testing.T) {
+			Reject:       true,
+		},
+
+		// Zones parameter specified do not match ReplicaCount [Fail]
```
it's technically "not enough" replicaCount

msau42 commented on pkg/volume/util/util_test.go:
```diff
@@ -1460,6 +1460,612 @@ func TestSelectZoneForVolume(t *testing.T) {
+		},
+
+		// Select zones from node label and AllowedTopologies [Pass]
+		// [1] nil Node
```
this comment is not correct

msau42 commented on pkg/volume/util/util_test.go:
```diff
@@ -1460,6 +1460,612 @@ func TestSelectZoneForVolume(t *testing.T) {
+			},
+			Reject:        false,
+			ExpectSpecificZones: true,
+			ExpectedZones: "zoneX,zoneY",
```
ExpectedZone

msau42 commented on pkg/volume/util/util_test.go:
```diff
@@ -1460,6 +1460,612 @@ func TestSelectZoneForVolume(t *testing.T) {
+		},
+
+		// Select zones from node label and AllowedTopologies [Pass]
+		// [1] nil Node
```
not correct

msau42 commented on pkg/volume/util/util_test.go:
```diff
@@ -1460,6 +1460,612 @@ func TestSelectZoneForVolume(t *testing.T) {
+			ReplicaCount:        2,
+			Reject:              false,
+			ExpectSpecificZones: true,
+			ExpectedZones:       "zoneW,zoneX",
```
I know ExpectedZone is redundant when ExpectSpecificZones is true, but may be good to explicitly call out anyway.

msau42 commented on pkg/volume/util/util_test.go:
```diff
@@ -1460,6 +1460,612 @@ func TestSelectZoneForVolume(t *testing.T) {
+			},
+			Reject:             false,
+			ExpectSpecificZones: true,
+			ExpectedZones:      "zoneX,zoneY",
```
ExpectedZone

msau42 commented:
/assign @verult

msau42 commented:
lgtm! will also let @verult review

msau42 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

ddebroy commented on pkg/volume/util/util.go:
```diff
@@ -320,58 +320,86 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
+			return sets.NewString(), fmt.Errorf("zone[s] cannot be specified in StorageClass if allowedTopologies specified")
 		}
-		return ChooseZoneForVolume(*allowedZones, pvcName), nil
+		return chooseZonesForVolumeIncludingZone(*allowedZones, pvcName, zoneFromNode, "allowedTopologies", numReplicas)
```
I clearly missed the fact that `nil` can be used to indicate an empty set. Thanks for point that above. With that, no dereferencing is necessary and I will fix that. 

I was using the label to be able to print out meaningful error messages without having to process the errors in the callers. However I am not sure that saves me much - so I will get rid of this label and error string inside `chooseZonesForVolumeIncludingZone` and process the errors in the callers.

ddebroy commented on pkg/volume/gce_pd/gce_util.go:
```diff
@@ -118,7 +118,10 @@ func (gceutil *GCEDiskUtil) CreateVolume(c *gcePersistentDiskProvisioner) (strin
-			configuredZones = v
+			configuredZones, err = volumeutil.ZonesToSet(v)
+			if err != nil {
+				return "", 0, nil, "", fmt.Errorf("error parsing zones %s, must be strings separated by commas: %v", v, err)
```
I have added the error to `ZonesToSet`. Will adjust the error message in other consumers (like EBS/Azure) in a separate PR.

ddebroy commented:
/test pull-kubernetes-integration

verult reviewed (commented):
Initial pass

msau42 commented on pkg/volume/gce_pd/gce_pd.go:
```diff
@@ -526,7 +527,17 @@ func (c *gcePersistentDiskProvisioner) Provision(selectedNode *v1.Node, allowedT
 			pv.Labels[k] = v
-			requirements = append(requirements, v1.NodeSelectorRequirement{Key: k, Operator: v1.NodeSelectorOpIn, Values: []string{v}})
+			var values []string
+			if k == kubeletapis.LabelZoneFailureDomain {
```
can this have a unit test?

ddebroy commented on pkg/volume/gce_pd/gce_pd.go:
```diff
@@ -519,14 +519,23 @@ func (c *gcePersistentDiskProvisioner) Provision(selectedNode *v1.Node, allowedT
 		}
 		for k, v := range labels {
 			pv.Labels[k] = v
+			requirements = append(requirements, v1.NodeSelectorRequirement{Key: k, Operator: v1.NodeSelectorOpIn, Values: []string{v}})
```
Thanks for catching this.

verult commented on pkg/volume/util/util.go:
```diff
@@ -320,62 +320,98 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
 	if zoneParameterPresent {
-		return zoneParameter, nil
+		if numReplicas > 1 {
+			return sets.NewString(), fmt.Errorf("zone cannot be specified if desired number of replicas for pv is greather than 1. Please specify zones or allowedTopologies to specify desired zones")
```
return nil to be consistent

verult commented on pkg/volume/util/util.go:
```diff
@@ -539,6 +579,25 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
+// chooseZonesForVolumeIncludingZone is a wrapper around ChooseZonesForVolume that ensures zoneToInclude is chosen
+func chooseZonesForVolumeIncludingZone(zones sets.String, pvcName, zoneToInclude string, numReplicas uint32) (sets.String, error) {
+	if uint32(zones.Len()) < numReplicas {
+		return sets.NewString(), fmt.Errorf("not enough zones found to provision a volume with %d replicas. Need at least %d distinct zones for a volume with %d replicas", numReplicas, numReplicas, numReplicas)
```
return `nil` to be consistent with the rest of `SelectZonesForVolume()`.

verult commented on pkg/volume/util/util.go:
```diff
@@ -539,6 +579,25 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
+	if uint32(zones.Len()) == numReplicas {
+		return zones, nil
+	}
+	if zoneToInclude != "" {
```
What if `zoneToInclude` is not in `zones`?

verult replied:
If this function assumes `zoneToInclude` is always in `zones`, it'd be helpful to call it out in comments

ddebroy replied:
The scheduler will make sure `zoneToInclude` in always in `zones`. Will call out the assumption explicitly.

verult replied:
Maybe you are referring to a different function? Both calls to `choozeZonesForVolumeIncludingZone()` pass in non-empty `zones` and they both allow for the possibility of non-empty `zoneToInclude`

ddebroy replied:
Sorry - I was indeed referring to a different function. Agree with your comment above - `choozeZonesForVolumeIncludingZone()` does expect `zoneToInclude` to be part of `zones` and I will clarify that in a comment.

verult commented on pkg/volume/util/util.go:
```diff
@@ -320,62 +320,98 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
-			return "", fmt.Errorf("zone[s] cannot be specified in StorageClass if allowedTopologies specified")
+			return nil, fmt.Errorf("zone[s] cannot be specified in StorageClass if allowedTopologies specified")
+		}
+		if zones, err := chooseZonesForVolumeIncludingZone(allowedZones, pvcName, zoneFromNode, numReplicas); err != nil {
```
Should we validate that `zoneFromNode` is inside `allowedZones` before this call?

verult commented on pkg/volume/util/util.go:
```diff
@@ -320,62 +320,98 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
+		}
+		// if single replica volume and node with zone found, return immediately
+		if numReplicas == 1 {
+			return sets.NewString(zoneFromNode), nil
```
`zoneFromNode` could be outside allowed zones

verult replied:
Sorry I was referring to `allowedTopologies`. Thanks for the explanation!

verult commented on pkg/volume/util/util.go:
```diff
@@ -320,62 +320,98 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
+// SelectZonesForVolume selects zones for a volume based on several factors:
+// node.zone, allowedTopologies, zone/zones parameters from storageclass,
+// zones with active nodes from the cluster. The number of zones = replicas.
+func SelectZonesForVolume(zoneParameterPresent, zonesParameterPresent bool, zoneParameter string, zonesParameter, zonesWithNodes sets.String, node *v1.Node, allowedTopologies []v1.TopologySelectorTerm, pvcName string, numReplicas uint32) (sets.String, error) {
```
nit: Is it possible to do without some of these parameters? If we can reduce the input set, it'll help guarantee test coverage and possibly improve code maintainability

davidz627 replied:
since a lot of these parameters are mutually exclusive to each other, it might be worth having a `SelectZonesForVolume` that only takes in one of these params called `zones`, and a wrapper function that does all this logic to determine mutual exclusivity as well as picking which of the `zone|zones|topology` to pass to `SelectZonesForVolume`

ddebroy replied:
@verult The idea behind `SelectZonesForVolume` is all the in-tree providers like EBS, Azure and GCE PD should be able to parse out volume zone and topology parameters from the storage class, invoke `SelectZoneForVolume`/`SelectZonesForVolume` and get a (set of) volume(s) back to pass to the cloud provider API without having to duplicate any of the logic to enforce the rules. So `SelectZonesForVolume` is the wrapper that implements all the common logic to determine mutual exclusivity and for that it needs the full set of parameters around zone/zones/node/topology. I agree that the long list of parameters is not pretty but to help with maintenance, we have an extensive set of unit tests in `TestSelectZonesForVolume` and `TestSelectZoneForVolume`. Those should make sure none of the rules around exclusivity of the parameters are broken.

@davidz627  `SelectZonesForVolume` is the wrapper that enforces the compatibility rules among the parameters. `chooseZonesForVolumeIncludingZone`/`ChooseZonesForVolume` invoked from `SelectZonesForVolume` does the actual picking of zone(s) from a set of zones. Let me know if it makes sense and already aligns with the above suggestion.

davidz627 commented on pkg/volume/gce_pd/gce_pd.go:
```diff
@@ -519,14 +520,33 @@ func (c *gcePersistentDiskProvisioner) Provision(selectedNode *v1.Node, allowedT
+			if k == kubeletapis.LabelZoneFailureDomain {
+				zones, err := util.LabelZonesToSet(v)
+				if err != nil {
+					return nil, fmt.Errorf("failed to convert label string for Zone: %s to a Set", v)
```
nit: wrap error here

davidz627 commented on pkg/volume/gce_pd/gce_pd.go:
```diff
@@ -519,14 +520,33 @@ func (c *gcePersistentDiskProvisioner) Provision(selectedNode *v1.Node, allowedT
+				if err != nil {
+					return nil, fmt.Errorf("failed to convert label string for Zone: %s to a Set", v)
+				}
+				values = zones.List()
```
can we go straight from label to list here instead of going through an intermediate set?

davidz627 commented on pkg/volume/util/util.go:
```diff
@@ -397,7 +433,11 @@ func ZonesSetToLabelValue(strSet sets.String) string {
+	if zones, err := stringToSet(zonesString, ","); err != nil {
+		return nil, fmt.Errorf("error parsing zones %s, must be strings separated by commas: %v", zonesString, err)
+	} else {
+		return zones, nil
```
nit: doesn't have to be in `else` block

ddebroy commented on pkg/volume/gce_pd/gce_util.go:
```diff
@@ -148,61 +147,44 @@ func (gceutil *GCEDiskUtil) CreateVolume(c *gcePersistentDiskProvisioner) (strin
+	var activezones sets.String
+	activezones, err = cloud.GetAllCurrentZones()
+	if err != nil {
+		return "", 0, nil, "", err
```
There was a review comment earlier to remove the log as the error already captures it. https://github.com/kubernetes/kubernetes/pull/67530#discussion_r210962582 . Do we really need it?

msau42 replied:
Generally, we should avoid logging error messages that we also return because that creates duplicate log messages.

verult replied:
ACK

ddebroy commented on pkg/volume/util/util.go:
```diff
@@ -320,62 +320,98 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
 	// pick zone from parameters if present
 	if zoneParameterPresent {
-		return zoneParameter, nil
+		if numReplicas > 1 {
```
Is it necessary? All `ValidateZone` is doing is making sure the string does not have spaces/tabs. The cloud API call for PD creation will fail for an invalid zone like that eventually. Maybe I can invoke `ValidateZone` on the list of returned zones from `SelectZonesForVolume` if it is important.

verult commented on pkg/volume/util/util.go:
```diff
@@ -320,62 +320,98 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
-		// pick node's zone and ignore any allowedTopologies (since scheduler is assumed to have accounted for it already)
-		z, ok := node.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
+		// pick node's zone for one of the replicas
+		zoneFromNode, ok := node.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
```
Is `zoneFromNode` only being set inside the scope of this block?

ddebroy replied:
Yes this is the only place `zoneFromNode` is set.

ddebroy commented on staging/src/k8s.io/apimachinery/pkg/apis/config/v1alpha1/BUILD:
```diff
@@ -10,6 +10,7 @@ go_library(
         "zz_generated.conversion.go",
         "zz_generated.deepcopy.go",
+        "zz_generated.defaults.go",
     ],
```
I performed a `update-bazel.sh` and that somehow caused it to pop up. Removed it and re-pushed to get rid of it.

ddebroy commented:
/remove-sig api-machinery

ddebroy commented:
@verult @msau42 @davidz627 I have addressed the review comments so far ... PTAL. 

Note that running update-bazel led to an unexpected change and that has since been removed from the PR above.

davidz627 commented:
@ddebroy sometimes update-bazel generates weird artifacts if you have a nested `kubernetes` folder inside your `kubernetes` folder, this gets autogenerated sometimes. You can try check for a `k8s.io/kubernetes/kubernetes` folder and just `rm -rf` or do a `git clean` to remove the nested one.

verult commented on pkg/volume/util/util.go:
```diff
@@ -539,6 +604,33 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
 
+// chooseZonesForVolumeIncludingZone is a wrapper around ChooseZonesForVolume that ensures zoneToInclude is chosen
+// zoneToInclude can either be empty in which case it is ignored. If non-empty, zoneToInclude is expected to be member of zones.
+// numReplicas is expected to be > 0 and >= zones.Len()
```
"<= zones.Len()"

davidz627 commented on pkg/volume/gce_pd/gce_pd_test.go:
```diff
@@ -91,6 +96,16 @@ func (fake *fakePDManager) DeleteVolume(cd *gcePersistentDiskDeleter) error {
 
+func getNodeSelectorRequirementWithKey(key string, term v1.NodeSelectorTerm) (*v1.NodeSelectorRequirement, error) {
+	for _, r := range term.MatchExpressions {
+		if r.Key != key {
```
nit: use positive expressions when possible, especially here this feels a little backwards

instead, maybe:
```
if r.Key == key{
  return &r, nil
}
```

davidz627 commented on pkg/volume/util/util.go:
```diff
@@ -320,62 +320,102 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
-		// pick node's zone and ignore any allowedTopologies (since scheduler is assumed to have accounted for it already)
-		z, ok := node.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
+		// pick node's zone for one of the replicas
+		ok := false
```
nit: `var ok bool`

davidz627 reviewed (commented):
final nits from me, lgtm otherwise!

ddebroy commented:
/test pull-kubernetes-node-e2e

davidz627 commented:
/lgtm

ddebroy commented:
/test pull-kubernetes-node-e2e

msau42 commented:
Can you squash your commits?

ddebroy committed:
Add DynamicProvisioningScheduling support for GCE PD and RePD

Signed-off-by: Deep Debroy <ddebroy@docker.com>

ddebroy commented:
squashed commits

saad-ali commented:
CC @verult 

Please ensure `func isRegionalPD(spec *volume.Spec) bool {` is updated in pkg/volume/gce_pd/gce_util.go to handle this new logic.

ddebroy commented:
@saad-ali `isRegionalPD` should continue to work as it used to from what I understand. There are no changes in how `LabelZoneFailureDomain` will get applied to PD/RePD due to this PR.

davidz627 commented:
/lgtm

ddebroy commented:
/test pull-kubernetes-e2e-kops-aws

verult commented:
@ddebroy yep it should still work, but since eventually we plan to use NodeAffinity as the main way to specify topology, `isRegionalPD()` should be able to determine if a spec is for regional PD based on NodeAffinity too. On the other hand, since it doesn't break existing functionality, I'm OK with tracking it as an issue and adding it in a separate PR.

saad-ali commented:
`isRegionalPD` will continue to work but it relies on the labels that this PR is replacing. We should update the method to use nodeAffinity (introduced in this PR) instead. But like @verult said we can do that in a follow up PR (he has a PR out touching that code, so we can do it there).

Otherwise LGTM.

/lgtm

msau42 commented:
For backwards compatibility, we cannot replace the labels check until the labels method is fully removed.

msau42 commented:
We intentionally keep the old labeling + NodeAffinity for this backwards compatibility reason.

verult commented:
@msau42 right, we could check using both NodeAffinity and labels, prioritizing NodeAffinity
